pinot-common/src/main/java/org/apache/pinot/common/function/scalar/DataTypeConversionFunctions.java

```diff
@@ -68,6 +69,14 @@ public static Object cast(Object value, String targetTypeLiteral) {
         }
         break;
     }
+    if (sourceType == STRING && targetDataType == LONG) {
+      // Check if the STRING is a TIMESTAMP
+      try {
+        return TimestampUtils.toTimestamp(value.toString().trim()).getTime();
+      } catch (Exception e) {
+        // If it is not a TIMESTAMP, we will try to parse it as a LONG, so do nothing here.
+      }
+    }
```

Comment:
@Jackie-Jiang I've intentionally not put this logic within `PinotDataType` since that's a little more widely used but let me know if you think otherwise. Also long term we might want to try to unify the casting logic between the scalar and transform function variants of `CAST`.

cc - @gortiz since you've been working on the UDF stuff.